{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install pycirclizely"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "from IPython.display import HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Circos plot: eukaryotes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqFeature import SeqFeature\n",
    "\n",
    "\n",
    "def hover_formatter(feature: SeqFeature) -> str:\n",
    "    \"\"\"Custom function for genomic features annotation hover text.\"\"\"\n",
    "    info = []\n",
    "\n",
    "    if hasattr(feature, \"qualifiers\") and feature.qualifiers.get(\"gene\", [\"\"])[0]:\n",
    "        info.append(f\"Gene: {feature.qualifiers['gene'][0]}\")\n",
    "\n",
    "    if hasattr(feature, \"type\") and feature.type:\n",
    "        info.append(f\"Type: {feature.type}\")\n",
    "\n",
    "    if hasattr(feature, \"location\") and hasattr(feature.location, \"strand\"):\n",
    "        strand = \"+\" if feature.location.strand == 1 else \"-\"\n",
    "        info.append(f\"Strand: {strand}\")\n",
    "\n",
    "    if hasattr(feature, \"location\"):\n",
    "        start = (\n",
    "            int(feature.location.start) if hasattr(feature.location, \"start\") else \"?\"\n",
    "        )\n",
    "        end = int(feature.location.end) if hasattr(feature.location, \"end\") else \"?\"\n",
    "        info.append(f\"Position: {start}-{end}\")\n",
    "\n",
    "    if hasattr(feature, \"qualifiers\") and feature.qualifiers.get(\"product\", [\"\"])[0]:\n",
    "        info.append(f\"Product: {feature.qualifiers['product'][0]}\")\n",
    "\n",
    "    return \"<br>\".join(info) if info else \"No information available\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4 Homo sapiens (hg38)\n",
    "\n",
    "**hg38** data files are obtained from [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables).\n",
    "\n",
    "Dataset Repository:  \n",
    "https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/hg38"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-1. Ideograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import load_eukaryote_example_dataset\n",
    "\n",
    "# Load hg38 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/hg38)\n",
    "chr_bed_file, cytoband_file, _ = load_eukaryote_example_dataset(\"hg38\")\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(chr_bed_file, space=3)\n",
    "circos.text(\"Homo sapiens (hg38)\", font=dict(size=15))\n",
    "\n",
    "# Add cytoband tracks from cytoband file\n",
    "circos.add_cytoband_tracks((95, 100), cytoband_file, show_hovertext=True)\n",
    "\n",
    "# Plot chromosome name\n",
    "for sector in circos.sectors:\n",
    "    sector.text(sector.name, font=dict(size=12))\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-2. Segmental Dups Link"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import ColorCycler, load_eukaryote_example_dataset\n",
    "\n",
    "# Load hg38 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/hg38)\n",
    "chr_bed_file, cytoband_file, chr_links = load_eukaryote_example_dataset(\"hg38\")\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(chr_bed_file, space=3)\n",
    "circos.text(\"Homo sapiens<br>(hg38)\", deg=315, r=150, font=dict(size=12))\n",
    "\n",
    "# Add cytoband tracks from cytoband file\n",
    "circos.add_cytoband_tracks((95, 100), cytoband_file)\n",
    "\n",
    "# Create chromosome color dict\n",
    "color_cycler = ColorCycler(\"Rainbow\")\n",
    "chr_names = [s.name for s in circos.sectors]\n",
    "colors = color_cycler.get_colors(len(chr_names))[::-1]\n",
    "chr_name2color = {name: color for name, color in zip(chr_names, colors)}\n",
    "\n",
    "# Plot chromosome name & xticks\n",
    "for sector in circos.sectors:\n",
    "    sector.text(\n",
    "        sector.name, r=120, font=dict(size=10, color=chr_name2color[sector.name])\n",
    "    )\n",
    "    sector.get_track(\"cytoband\").xticks_by_interval(\n",
    "        40000000,\n",
    "        label_orientation=\"vertical\",\n",
    "        label_formatter=lambda v: f\"{v / 1000000:.0f} Mb\",\n",
    "        text_kws=dict(font=dict(size=8)),\n",
    "    )\n",
    "\n",
    "# Plot chromosome link\n",
    "for link in chr_links:\n",
    "    region1 = (link.query_chr, link.query_start, link.query_end)\n",
    "    region2 = (link.ref_chr, link.ref_start, link.ref_end)\n",
    "    color = chr_name2color[link.query_chr]\n",
    "    if link.query_chr in (\"chr1\", \"chr8\", \"chr16\") and link.query_chr != link.ref_chr:\n",
    "        circos.link(\n",
    "            region1, region2, fillcolor=color, line=dict(color=color, width=0.1)\n",
    "        )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-3. Plot Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import ColorCycler, load_eukaryote_example_dataset\n",
    "\n",
    "np.random.seed(0)\n",
    "\n",
    "# Load hg38 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/hg38)\n",
    "chr_bed_file = load_eukaryote_example_dataset(\"hg38\")[0]\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(\n",
    "    chr_bed_file, start=-80, end=260, space=3, endspace=False\n",
    ")\n",
    "circos.text(\"Homo sapiens<br>(hg38)\", font=dict(size=15))\n",
    "\n",
    "# Create chromosome color dict\n",
    "color_cycler = ColorCycler(\"Rainbow\")\n",
    "chr_names = [s.name for s in circos.sectors]\n",
    "colors = color_cycler.get_colors(len(chr_names))[::-1]\n",
    "chr_name2color = {name: color for name, color in zip(chr_names, colors)}\n",
    "\n",
    "for sector in circos.sectors:\n",
    "    # Plot chromosome outer track\n",
    "    sector.text(sector.name.replace(\"chr\", \"\"))\n",
    "    color = chr_name2color[sector.name]\n",
    "    outer_track = sector.add_track((95, 100))\n",
    "    outer_track.axis(fillcolor=color)\n",
    "    # Create example x,y plot data\n",
    "    step = 10000000\n",
    "    x = np.arange(sector.start + (step / 2), sector.end - (step / 2), step)\n",
    "    y = np.random.randint(0, 100, size=len(x))\n",
    "    # Scatter track\n",
    "    track1 = sector.add_track((80, 90), r_pad_ratio=0.1)\n",
    "    track1.axis()\n",
    "    track1.scatter(x, y, vmax=100, marker=dict(size=3, color=\"red\"))\n",
    "    # Line track\n",
    "    track2 = sector.add_track((65, 75), r_pad_ratio=0.1)\n",
    "    track2.axis()\n",
    "    track2.line(x, y, vmax=100, line=dict(color=\"blue\"))\n",
    "    # Bar track\n",
    "    track3 = sector.add_track((50, 60), r_pad_ratio=0.1)\n",
    "    track3.axis()\n",
    "    track3.bar(x, y, vmax=100, width=step * 0.7, fillcolor=\"olive\")\n",
    "    # Fill between track\n",
    "    track4 = sector.add_track((35, 45), r_pad_ratio=0.1)\n",
    "    track4.axis()\n",
    "    track4.grid()\n",
    "    track4.fill_between(x, y, vmax=100, fillcolor=\"violet\")\n",
    "    # Plot track labels\n",
    "    if sector.name == circos.sectors[0].name:\n",
    "        circos.text(\"Chr\", r=outer_track.r_center, deg=-90)\n",
    "        circos.text(\"a\", r=track1.r_center, deg=-90, font=dict(color=\"red\"))\n",
    "        circos.text(\"b\", r=track2.r_center, deg=-90, font=dict(color=\"blue\"))\n",
    "        circos.text(\"c\", r=track3.r_center, deg=-90, font=dict(color=\"olive\"))\n",
    "        circos.text(\"d\", r=track4.r_center, deg=-90, font=dict(color=\"violet\"))\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5 Mus musculus (mm10)\n",
    "\n",
    "**mm10** data files are obtained from [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables).\n",
    "\n",
    "Dataset Repository:  \n",
    "https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/mm10"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-1. Ideograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import load_eukaryote_example_dataset\n",
    "\n",
    "# Load mm10 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/mm10)\n",
    "chr_bed_file, cytoband_file, _ = load_eukaryote_example_dataset(\"mm10\")\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(chr_bed_file, space=3)\n",
    "circos.text(\"Mus musculus (mm10)\", font=dict(size=15))\n",
    "\n",
    "# Add cytoband tracks from cytoband file\n",
    "circos.add_cytoband_tracks((95, 100), cytoband_file, show_hovertext=True)\n",
    "\n",
    "# Plot chromosome name\n",
    "for sector in circos.sectors:\n",
    "    sector.text(sector.name, font=dict(size=10))\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-2. Segmental Dups Link"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import ColorCycler, load_eukaryote_example_dataset\n",
    "\n",
    "# Load mm10 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/mm10)\n",
    "chr_bed_file, cytoband_file, chr_links = load_eukaryote_example_dataset(\"mm10\")\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(chr_bed_file, space=3)\n",
    "circos.text(\"Mus musculus<br>(mm10)\", deg=315, r=150, font=dict(size=12))\n",
    "\n",
    "# Add cytoband tracks from cytoband file\n",
    "circos.add_cytoband_tracks((95, 100), cytoband_file)\n",
    "\n",
    "# Create chromosome color dict\n",
    "color_cycler = ColorCycler(\"Rainbow\")\n",
    "chr_names = [s.name for s in circos.sectors]\n",
    "colors = color_cycler.get_colors(len(chr_names))[::-1]\n",
    "chr_name2color = {name: color for name, color in zip(chr_names, colors)}\n",
    "\n",
    "# Plot chromosome name & xticks\n",
    "for sector in circos.sectors:\n",
    "    sector.text(\n",
    "        sector.name, r=120, font=dict(size=10, color=chr_name2color[sector.name])\n",
    "    )\n",
    "    sector.get_track(\"cytoband\").xticks_by_interval(\n",
    "        50000000,\n",
    "        label_orientation=\"vertical\",\n",
    "        label_formatter=lambda v: f\"{v / 1000000:.0f} Mb\",\n",
    "        text_kws=dict(font=dict(size=8)),\n",
    "    )\n",
    "\n",
    "# Plot chromosome link\n",
    "for link in chr_links:\n",
    "    region1 = (link.query_chr, link.query_start, link.query_end)\n",
    "    region2 = (link.ref_chr, link.ref_start, link.ref_end)\n",
    "    color = chr_name2color[link.query_chr]\n",
    "    if link.query_chr != link.ref_chr:\n",
    "        circos.link(\n",
    "            region1, region2, fillcolor=color, line=dict(color=color, width=0.1)\n",
    "        )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-3. Plot Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import ColorCycler, load_eukaryote_example_dataset\n",
    "\n",
    "np.random.seed(0)\n",
    "\n",
    "# Load mm10 dataset (https://github.com/moshi4/pycirclizely-data/tree/main/eukaryote/mm10)\n",
    "chr_bed_file = load_eukaryote_example_dataset(\"mm10\")[0]\n",
    "\n",
    "# Initialize Circos from BED chromosomes\n",
    "circos = Circos.initialize_from_bed(\n",
    "    chr_bed_file, space=2, start=10, end=350, endspace=False\n",
    ")\n",
    "circos.text(\"Mus musculus<br>(mm10)\", font=dict(size=16))\n",
    "\n",
    "heatmap_cmaps = [\n",
    "    \"RdBu_r\",\n",
    "    \"Spectral\",\n",
    "    \"Viridis\",\n",
    "    \"Plasma\",\n",
    "    \"Reds\",\n",
    "    \"Blues\",\n",
    "    \"Greens\",\n",
    "    \"Greys\",\n",
    "]\n",
    "color_cycler = ColorCycler(\"Rainbow\")\n",
    "colors = color_cycler.get_colors(count=len(circos.sectors))[::-1]\n",
    "for sector, color in zip(circos.sectors, colors):\n",
    "    sector.text(sector.name.replace(\"chr\", \"\"), font=dict(size=12))\n",
    "    # Plot chromosome outer track\n",
    "    chr_track = sector.add_track((95, 100))\n",
    "    chr_track.axis(fillcolor=color)\n",
    "    # Create random test data for heatmap plot\n",
    "    window_size = 10_000_000\n",
    "    data_num = int(sector.size // window_size) + 1\n",
    "    vmin, vmax = 0, 100\n",
    "    data = np.random.randint(vmin, vmax, data_num)\n",
    "    # Plot heatmap tracks with various cmap\n",
    "    track_r_size, r_interval, r_start = 4, 2, 90\n",
    "    for idx, cmap in enumerate(heatmap_cmaps):\n",
    "        r_pos = r_start - (track_r_size + r_interval) * idx\n",
    "        track_r_lim = (r_pos - track_r_size, r_pos)\n",
    "        track = sector.add_track(track_r_lim)\n",
    "        track.axis(line=dict(color=\"grey\"))\n",
    "        track.heatmap(\n",
    "            data,\n",
    "            width=window_size,\n",
    "            vmin=vmin,\n",
    "            vmax=vmax,\n",
    "            cmap=cmap,\n",
    "            hover_text=None,\n",
    "            rect_kws=dict(line=dict(width=0.0)),\n",
    "        )\n",
    "        # Plot colormap name on center\n",
    "        if sector.name == circos.sectors[0].name:\n",
    "            circos.text(cmap, r=track.r_center, font=dict(size=10))\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyCirclizely",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
