{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install pycirclizely"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "from IPython.display import HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Circos plot: virus and prokaryotes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqFeature import SeqFeature\n",
    "\n",
    "\n",
    "def hover_formatter(feature: SeqFeature) -> str:\n",
    "    \"\"\"Custom function for genomic features annotation hover text.\"\"\"\n",
    "    info = []\n",
    "\n",
    "    if hasattr(feature, \"qualifiers\") and feature.qualifiers.get(\"gene\", [\"\"])[0]:\n",
    "        info.append(f\"Gene: {feature.qualifiers['gene'][0]}\")\n",
    "\n",
    "    if hasattr(feature, \"type\") and feature.type:\n",
    "        info.append(f\"Type: {feature.type}\")\n",
    "\n",
    "    if hasattr(feature, \"location\") and hasattr(feature.location, \"strand\"):\n",
    "        strand = \"+\" if feature.location.strand == 1 else \"-\"\n",
    "        info.append(f\"Strand: {strand}\")\n",
    "\n",
    "    if hasattr(feature, \"location\"):\n",
    "        start = (\n",
    "            int(feature.location.start) if hasattr(feature.location, \"start\") else \"?\"\n",
    "        )\n",
    "        end = int(feature.location.end) if hasattr(feature.location, \"end\") else \"?\"\n",
    "        info.append(f\"Position: {start}-{end}\")\n",
    "\n",
    "    if hasattr(feature, \"qualifiers\") and feature.qualifiers.get(\"product\", [\"\"])[0]:\n",
    "        info.append(f\"Product: {feature.qualifiers['product'][0]}\")\n",
    "\n",
    "    return \"<br>\".join(info) if info else \"No information available\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Enterobacteria phage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.parser import Gff\n",
    "from pycirclizely.utils import load_prokaryote_example_file\n",
    "\n",
    "# Load GFF file\n",
    "gff_file = load_prokaryote_example_file(\"enterobacteria_phage.gff\")\n",
    "gff = Gff(gff_file)\n",
    "\n",
    "# Initialize circos instance\n",
    "seqid2size = gff.get_seqid2size()\n",
    "space = 0 if len(seqid2size) == 1 else 2\n",
    "circos = Circos(sectors=seqid2size, space=space)\n",
    "circos.text(\"Enterobacteria phage<br>(NC_000902)\", font=dict(size=20))\n",
    "\n",
    "seqid2features = gff.get_seqid2features(feature_type=\"CDS\")\n",
    "for sector in circos.sectors:\n",
    "    cds_track = sector.add_track((90, 100))\n",
    "    cds_track.axis(fillcolor=\"#EEEEEE\", line=dict(width=0))\n",
    "\n",
    "    features = seqid2features[sector.name]\n",
    "    for feature in features:\n",
    "        # Plot CDS features\n",
    "        if feature.location.strand == 1:\n",
    "            cds_track.genomic_features(\n",
    "                feature,\n",
    "                plotstyle=\"arrow\",\n",
    "                r_lim=(95, 100),\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"salmon\",\n",
    "            )\n",
    "        else:\n",
    "            cds_track.genomic_features(\n",
    "                feature,\n",
    "                plotstyle=\"arrow\",\n",
    "                r_lim=(90, 95),\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"skyblue\",\n",
    "            )\n",
    "\n",
    "    # Plot xticks & intervals on inner position\n",
    "    cds_track.xticks_by_interval(\n",
    "        interval=5000,\n",
    "        outer=False,\n",
    "        label_formatter=lambda v: f\"{v/ 1000:.1f} Kb\",\n",
    "        label_orientation=\"vertical\",\n",
    "        line_kws=dict(line=dict(color=\"grey\")),\n",
    "    )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Escherichia coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from pycirclizely import Circos\n",
    "from pycirclizely.parser import Genbank\n",
    "from pycirclizely.utils import load_prokaryote_example_file\n",
    "\n",
    "# Load Genbank file\n",
    "gbk_file = load_prokaryote_example_file(\"escherichia_coli.gbk.gz\")\n",
    "gbk = Genbank(gbk_file)\n",
    "\n",
    "# Initialize circos instance\n",
    "seqid2size = gbk.get_seqid2size()\n",
    "space = 0 if len(seqid2size) == 1 else 2\n",
    "circos = Circos(sectors=seqid2size, space=space)\n",
    "circos.text(\"Escherichia coli<br>(NC_000913)\", font=dict(size=14))\n",
    "\n",
    "seqid2features = gbk.get_seqid2features(feature_type=None)\n",
    "seqid2seq = gbk.get_seqid2seq()\n",
    "for sector in circos.sectors:\n",
    "    # Plot outer track with xticks\n",
    "    major_ticks_interval = 500000\n",
    "    minor_ticks_interval = 100000\n",
    "    outer_track = sector.add_track((98, 100))\n",
    "    outer_track.axis(fillcolor=\"lightgrey\")\n",
    "    if sector.size >= major_ticks_interval:\n",
    "        outer_track.xticks_by_interval(\n",
    "            major_ticks_interval, label_formatter=lambda v: f\"{v/ 10 ** 6:.1f} Mb\"\n",
    "        )\n",
    "        outer_track.xticks_by_interval(\n",
    "            minor_ticks_interval, tick_length=1, show_label=False\n",
    "        )\n",
    "\n",
    "    f_cds_track = sector.add_track((90, 97), r_pad_ratio=0.1)\n",
    "    r_cds_track = sector.add_track((83, 90), r_pad_ratio=0.1)\n",
    "    rrna_track = sector.add_track((76, 83), r_pad_ratio=0.1)\n",
    "    trna_track = sector.add_track((69, 76), r_pad_ratio=0.1)\n",
    "\n",
    "    # Plot Forward CDS, Reverse CDS, rRNA, tRNA\n",
    "    features = seqid2features[sector.name]\n",
    "    for feature in features:\n",
    "        if feature.type == \"CDS\" and feature.location.strand == 1:\n",
    "            f_cds_track.genomic_features(\n",
    "                feature,\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"red\",\n",
    "                line=dict(color=None, width=0),\n",
    "            )\n",
    "        elif feature.type == \"CDS\" and feature.location.strand == -1:\n",
    "            r_cds_track.genomic_features(\n",
    "                feature,\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"blue\",\n",
    "                line=dict(color=None, width=0),\n",
    "            )\n",
    "        elif feature.type == \"rRNA\":\n",
    "            rrna_track.genomic_features(\n",
    "                feature,\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"green\",\n",
    "                line=dict(color=None, width=0),\n",
    "            )\n",
    "        elif feature.type == \"tRNA\":\n",
    "            trna_track.genomic_features(\n",
    "                feature,\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                line=dict(color=\"magenta\", width=0.1),\n",
    "            )\n",
    "\n",
    "    # Plot GC content\n",
    "    gc_content_track = sector.add_track((50, 65))\n",
    "    seq = seqid2seq[sector.name]\n",
    "    label_pos_list, gc_contents = gbk.calc_gc_content(seq=seq)\n",
    "    gc_contents = gc_contents - gbk.calc_genome_gc_content(seq=gbk.full_genome_seq)\n",
    "    positive_gc_contents = np.where(gc_contents > 0, gc_contents, 0)\n",
    "    negative_gc_contents = np.where(gc_contents < 0, gc_contents, 0)\n",
    "    abs_max_gc_content = np.max(np.abs(gc_contents))\n",
    "    vmin, vmax = -abs_max_gc_content, abs_max_gc_content\n",
    "    gc_content_track.fill_between(\n",
    "        label_pos_list, positive_gc_contents, 0, vmin=vmin, vmax=vmax, fillcolor=\"black\"\n",
    "    )\n",
    "    gc_content_track.fill_between(\n",
    "        label_pos_list, negative_gc_contents, 0, vmin=vmin, vmax=vmax, fillcolor=\"grey\"\n",
    "    )\n",
    "\n",
    "    # Plot GC skew\n",
    "    gc_skew_track = sector.add_track((35, 50))\n",
    "    label_pos_list, gc_skews = gbk.calc_gc_skew(seq=seq)\n",
    "    positive_gc_skews = np.where(gc_skews > 0, gc_skews, 0)\n",
    "    negative_gc_skews = np.where(gc_skews < 0, gc_skews, 0)\n",
    "    abs_max_gc_skew = np.max(np.abs(gc_skews))\n",
    "    vmin, vmax = -abs_max_gc_skew, abs_max_gc_skew\n",
    "    gc_skew_track.fill_between(\n",
    "        label_pos_list, positive_gc_skews, 0, vmin=vmin, vmax=vmax, fillcolor=\"olive\"\n",
    "    )\n",
    "    gc_skew_track.fill_between(\n",
    "        label_pos_list, negative_gc_skews, 0, vmin=vmin, vmax=vmax, fillcolor=\"purple\"\n",
    "    )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "# Add legend using dummy traces\n",
    "legend_labels = [\n",
    "    \"Forward CDS\",\n",
    "    \"Reverse CDS\",\n",
    "    \"rRNA\",\n",
    "    \"tRNA\",\n",
    "    \"Positive GC Content\",\n",
    "    \"Negative GC Content\",\n",
    "    \"Positive GC Skew\",\n",
    "    \"Negative GC Skew\",\n",
    "]\n",
    "\n",
    "legend_colors = [\"red\", \"blue\", \"green\", \"magenta\", \"black\", \"grey\", \"olive\", \"purple\"]\n",
    "\n",
    "marker_symbols = [\n",
    "    \"square\",\n",
    "    \"square\",\n",
    "    \"square\",\n",
    "    \"square\",\n",
    "    \"triangle-up\",\n",
    "    \"triangle-down\",\n",
    "    \"triangle-up\",\n",
    "    \"triangle-down\",\n",
    "]\n",
    "\n",
    "# Add legend with dummy trace\n",
    "for label, color, symbol in zip(legend_labels, legend_colors, marker_symbols):\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[None],\n",
    "            y=None,\n",
    "            mode=\"markers\",\n",
    "            marker=dict(color=color, symbol=symbol, size=9),\n",
    "            name=label,\n",
    "            showlegend=True,\n",
    "        )\n",
    "    )\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        x=1,\n",
    "        y=0.05,\n",
    "        xanchor=\"center\",\n",
    "        yanchor=\"middle\",\n",
    "        font=dict(size=12),\n",
    "        bgcolor=\"rgba(255,255,255,0.0)\",\n",
    "    )\n",
    ")\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Mycoplasma alvi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclizely import Circos\n",
    "from pycirclizely.parser import Genbank, Gff\n",
    "from pycirclizely.utils import load_prokaryote_example_file\n",
    "\n",
    "# Case1. Load `GFF` contig genomes\n",
    "# https://github.com/moshi4/pycirclizely-data/blob/main/prokaryote/mycoplasma_alvi.gff\n",
    "gff_file = load_prokaryote_example_file(\"mycoplasma_alvi.gff\")\n",
    "gff_parser = Gff(gff_file)\n",
    "\n",
    "# Case2. Load `Genbank` contig genomes (try this one)\n",
    "# https://github.com/moshi4/pycirclizely-data/blob/main/prokaryote/mycoplasma_alvi.gbk\n",
    "gbk_file = load_prokaryote_example_file(\"mycoplasma_alvi.gbk\")\n",
    "gbk_parser = Genbank(gbk_file)\n",
    "\n",
    "# Initialize circos instance\n",
    "seqid2size = gbk_parser.get_seqid2size()\n",
    "space = 0 if len(seqid2size) == 1 else 2\n",
    "circos = Circos(seqid2size, space=space)\n",
    "circos.text(\n",
    "    f\"Mycoplasma alvi<br>({len(circos.sectors)} contigs)\", r=15, font=dict(size=20)\n",
    ")\n",
    "\n",
    "seqid2features = gbk_parser.get_seqid2features(feature_type=None)\n",
    "for sector in circos.sectors:\n",
    "    # Plot outer track\n",
    "    outer_track = sector.add_track((99.7, 100))\n",
    "    outer_track.axis(fillcolor=\"black\")\n",
    "    major_interval = 100000\n",
    "    minor_interval = int(major_interval / 10)\n",
    "    if sector.size > minor_interval:\n",
    "        outer_track.xticks_by_interval(\n",
    "            major_interval, label_formatter=lambda v: f\"{v / 1000:.0f} Kb\"\n",
    "        )\n",
    "        outer_track.xticks_by_interval(minor_interval, tick_length=1, show_label=False)\n",
    "\n",
    "    # Plot forward/reverse CDS, rRNA, tRNA tracks\n",
    "    f_cds_track = sector.add_track((91, 98), r_pad_ratio=0.1)\n",
    "    r_cds_track = sector.add_track((84, 91), r_pad_ratio=0.1)\n",
    "    rrna_track = sector.add_track((77, 84), r_pad_ratio=0.1)\n",
    "    trna_track = sector.add_track((70, 77), r_pad_ratio=0.1)\n",
    "    for feature in seqid2features[sector.name]:\n",
    "        if feature.type == \"CDS\":\n",
    "            if feature.location.strand == 1:\n",
    "                f_cds_track.genomic_features(\n",
    "                    [feature],\n",
    "                    hover_text_formatter=hover_formatter,\n",
    "                    fillcolor=\"tomato\",\n",
    "                    line=dict(width=0.0),\n",
    "                )\n",
    "            else:\n",
    "                r_cds_track.genomic_features(\n",
    "                    [feature],\n",
    "                    hover_text_formatter=hover_formatter,\n",
    "                    fillcolor=\"skyblue\",\n",
    "                    line=dict(width=0.0),\n",
    "                )\n",
    "        elif feature.type == \"rRNA\":\n",
    "            rrna_track.genomic_features(\n",
    "                [feature],\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"lime\",\n",
    "                line=dict(width=0.0),\n",
    "            )\n",
    "        elif feature.type == \"tRNA\":\n",
    "            trna_track.genomic_features(\n",
    "                [feature],\n",
    "                hover_text_formatter=hover_formatter,\n",
    "                fillcolor=\"magenta\",\n",
    "                line=dict(width=0.0),\n",
    "            )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "\n",
    "# Add legend using dummy traces\n",
    "legend_labels = [\n",
    "    \"Forward CDS\",\n",
    "    \"Reverse CDS\",\n",
    "    \"rRNA\",\n",
    "    \"tRNA\",\n",
    "]\n",
    "\n",
    "legend_colors = [\"red\", \"blue\", \"green\", \"magenta\"]\n",
    "\n",
    "# Add all legend entries in a loop\n",
    "for label, color in zip(legend_labels, legend_colors):\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[None],\n",
    "            y=None,\n",
    "            mode=\"markers\",\n",
    "            marker=dict(color=color, symbol=\"square\", size=12),\n",
    "            name=label,\n",
    "            showlegend=True,\n",
    "        )\n",
    "    )\n",
    "\n",
    "# Update layout with legend position and style\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=0.40,\n",
    "        xanchor=\"center\",\n",
    "        yanchor=\"middle\",\n",
    "        font=dict(size=12),\n",
    "        bgcolor=\"rgba(255,255,255,0.0)\",\n",
    "    )\n",
    ")\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyCirclizely",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
