{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install pycirclizely pygenomeviz\n",
    "# !apt install ncbi-blast+ mummer mmseqs2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "from IPython.display import HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Advanced users can plot figures for comparative genomics flexibly with pyCirclize API.\n",
    "In this notebook, simple code recipes for comparative genomics Circos visualization utilizing [pyGenomeViz](https://github.com/moshi4/pyGenomeViz) align module are shown.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MUMmer\n",
    "\n",
    "Plot MUMmer alignment links between query-reference genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pygenomeviz.align import MUMmer  # type: ignore[import-untyped]\n",
    "from pygenomeviz.parser import Genbank  # type: ignore[import-untyped]\n",
    "from pygenomeviz.utils import (  # type: ignore[import-untyped]\n",
    "    load_example_genbank_dataset,\n",
    ")\n",
    "\n",
    "from pycirclizely import Circos\n",
    "\n",
    "TICKS_INTERVAL = 1000000\n",
    "\n",
    "# Load query & reference genbank files\n",
    "gbk_files = load_example_genbank_dataset(\"escherichia_coli\")\n",
    "ref_gbk = Genbank(gbk_files[2])\n",
    "query_gbk = Genbank(gbk_files[3])\n",
    "\n",
    "# Initialize circos instance\n",
    "circos = Circos(\n",
    "    sectors=dict(\n",
    "        **ref_gbk.get_seqid2size(),\n",
    "        **dict(reversed(list(query_gbk.get_seqid2size().items()))),\n",
    "    ),\n",
    "    start=-358,\n",
    "    end=2,\n",
    "    space=4,\n",
    "    sector2clockwise={seqid: False for seqid in query_gbk.get_seqid2size().keys()},\n",
    ")\n",
    "circos.text(\n",
    "    f\"{ref_gbk.name}\\n({ref_gbk.full_genome_length:,} bp)\",\n",
    "    r=130,\n",
    "    deg=35,\n",
    "    font=dict(size=13),\n",
    ")\n",
    "circos.text(\n",
    "    f\"{query_gbk.name}\\n({query_gbk.full_genome_length:,} bp)\",\n",
    "    r=130,\n",
    "    deg=-35,\n",
    "    font=dict(size=13),\n",
    ")\n",
    "\n",
    "# Plot genomic sector axis & xticks\n",
    "for sector in circos.sectors:\n",
    "    track = sector.add_track((99.8, 100))\n",
    "    track.axis(fillcolor=\"black\")\n",
    "    if sector.size >= TICKS_INTERVAL:\n",
    "        track.xticks_by_interval(\n",
    "            TICKS_INTERVAL,\n",
    "            label_formatter=lambda v: f\"{v/1000000:.1f} Mb\",\n",
    "            label_orientation=\"vertical\",\n",
    "        )\n",
    "\n",
    "# MUMmer genome comparison & plot links\n",
    "align_coords = MUMmer([query_gbk, ref_gbk]).run()\n",
    "for ac in align_coords:\n",
    "    region1 = (ac.query_name, ac.query_start, ac.query_end)\n",
    "    region2 = (ac.ref_name, ac.ref_start, ac.ref_end)\n",
    "    color = \"red\" if ac.is_inverted else \"grey\"\n",
    "    circos.link(region1, region2, fillcolor=color, line=dict(color=color), r1=98, r2=98)\n",
    "\n",
    "fig = circos.plotfig()\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MMseqs\n",
    "\n",
    "Search RBH(Reciprocal Best Hit) CDSs and plot hit links between query-reference genomes by MMseqs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from pygenomeviz.align import MMseqs  # type: ignore[import-untyped]\n",
    "from pygenomeviz.parser import Genbank  # type: ignore[import-untyped]\n",
    "from pygenomeviz.utils import (  # type: ignore[import-untyped]\n",
    "    load_example_genbank_dataset,\n",
    ")\n",
    "\n",
    "from pycirclizely import Circos\n",
    "\n",
    "QUERY_COLOR = \"salmon\"\n",
    "REF_COLOR = \"skyblue\"\n",
    "TICKS_INTERVAL = 100000\n",
    "\n",
    "# Load query & reference genbank files\n",
    "gbk_files = load_example_genbank_dataset(\"mycoplasma_mycoides\")\n",
    "query_gbk = Genbank(gbk_files[3])\n",
    "ref_gbk = Genbank(gbk_files[2])\n",
    "\n",
    "# Initialize circos instance\n",
    "circos = Circos(\n",
    "    sectors=dict(\n",
    "        **ref_gbk.get_seqid2size(),\n",
    "        **dict(reversed(list(query_gbk.get_seqid2size().items()))),\n",
    "    ),\n",
    "    start=-358,\n",
    "    end=2,\n",
    "    space=4,\n",
    "    sector2clockwise={seqid: False for seqid in query_gbk.get_seqid2size()},\n",
    ")\n",
    "\n",
    "# Plot genomic sector axis & xticks\n",
    "for sector in circos.sectors:\n",
    "    track = sector.add_track((98, 100))\n",
    "    color = QUERY_COLOR if sector.name in query_gbk.get_seqid2size() else REF_COLOR\n",
    "    track.axis(fillcolor=color)\n",
    "    if sector.size >= TICKS_INTERVAL:\n",
    "        track.xticks_by_interval(\n",
    "            TICKS_INTERVAL,\n",
    "            label_formatter=lambda v: f\"{v/1000000:.1f} Mb\",\n",
    "            label_orientation=\"vertical\",\n",
    "        )\n",
    "\n",
    "# Search MMseqs RBH CDSs & plot hit links\n",
    "align_coords = MMseqs([query_gbk, ref_gbk]).run()\n",
    "for ac in align_coords:\n",
    "    region1 = (ac.query_name, ac.query_start, ac.query_end)\n",
    "    region2 = (ac.ref_name, ac.ref_start, ac.ref_end)\n",
    "    color = \"red\" if ac.is_inverted else \"grey\"\n",
    "    circos.link(region1, region2, fillcolor=color, line=dict(color=color))\n",
    "\n",
    "fig = circos.plotfig()\n",
    "\n",
    "# Add legend using dummy traces\n",
    "legend_labels = [\n",
    "    f\"{query_gbk.name} (Query)\",\n",
    "    f\"{ref_gbk.name} (Ref)\",\n",
    "]\n",
    "\n",
    "legend_colors = [QUERY_COLOR, REF_COLOR]\n",
    "\n",
    "for label, color in zip(legend_labels, legend_colors):\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[None],\n",
    "            y=None,\n",
    "            mode=\"markers\",\n",
    "            marker=dict(color=color, symbol=\"square\", size=12),\n",
    "            name=label,\n",
    "            showlegend=True,\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=-0.1,\n",
    "        xanchor=\"center\",\n",
    "        yanchor=\"top\",\n",
    "        font=dict(size=12),\n",
    "        bgcolor=\"rgba(255,255,255,0.5)\",\n",
    "        orientation=\"h\",\n",
    "    )\n",
    ")\n",
    "\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blast\n",
    "\n",
    "Plot [BRIG](https://doi.org/10.1186/1471-2164-12-402)(Blast Ring Image Generator) like genome comparison figure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from pygenomeviz.align import AlignCoord, Blast  # type: ignore[import-untyped]\n",
    "from pygenomeviz.parser import Fasta  # type: ignore[import-untyped]\n",
    "from pygenomeviz.utils import (  # type: ignore[import-untyped]\n",
    "    ColorCycler,\n",
    "    interpolate_color,\n",
    "    load_example_fasta_dataset,\n",
    ")\n",
    "\n",
    "from pycirclizely import Circos\n",
    "from pycirclizely.utils import parse_color\n",
    "\n",
    "ColorCycler.set_cmap(\"Set1\")\n",
    "\n",
    "QUERY_TRACK_SIZE = 5\n",
    "MIN_IDENTITY = 70\n",
    "TICKS_INTERVAL = 100000\n",
    "\n",
    "# Load target & comparison fasta files\n",
    "fasta_files = load_example_fasta_dataset(\"mycoplasma_mycoides\")\n",
    "target_fasta = Fasta(fasta_files[0])\n",
    "comp_fasta_list = list(map(Fasta, fasta_files[1:]))\n",
    "\n",
    "# Initialize circos instance\n",
    "circos = Circos(\n",
    "    sectors=target_fasta.get_seqid2size(),\n",
    "    space=0 if len(target_fasta.get_seqid2size()) == 1 else 2,\n",
    ")\n",
    "circos.text(\n",
    "    f\"{target_fasta.name}<br>({target_fasta.full_genome_length:,} bp)\",\n",
    "    font=dict(size=16),\n",
    ")\n",
    "\n",
    "# Blast genome comparison & plot match blocks\n",
    "min_r_pos = 100\n",
    "comp_name2color = {}\n",
    "for idx, comp_fasta in enumerate(comp_fasta_list):\n",
    "    align_coords = Blast([target_fasta, comp_fasta]).run()\n",
    "    align_coords = AlignCoord.filter(align_coords, identity_thr=MIN_IDENTITY)\n",
    "    color = ColorCycler()\n",
    "    comp_name2color[comp_fasta.name] = color\n",
    "    min_r_pos -= QUERY_TRACK_SIZE\n",
    "    for sector in circos.sectors:\n",
    "        sector.add_track((min_r_pos, min_r_pos + QUERY_TRACK_SIZE), r_pad_ratio=0.1)\n",
    "    for ac in align_coords:\n",
    "        # Last added track in sector\n",
    "        track = circos.get_sector(ac.query_name).tracks[-1]\n",
    "        rect_color = interpolate_color(color, v=ac.identity, vmin=MIN_IDENTITY)\n",
    "        track.rect(\n",
    "            ac.query_start, ac.query_end, fillcolor=rect_color, line=dict(width=0)\n",
    "        )\n",
    "\n",
    "# Plot genomic sector axis & xticks\n",
    "for sector in circos.sectors:\n",
    "    track = sector.add_track((min_r_pos - 0.3, min_r_pos))\n",
    "    track.axis(fillcolor=\"black\")\n",
    "    if sector.size >= TICKS_INTERVAL:\n",
    "        track.xticks_by_interval(\n",
    "            TICKS_INTERVAL,\n",
    "            outer=False,\n",
    "            label_formatter=lambda v: f\"{v/1000000:.1f} Mb\",\n",
    "            label_orientation=\"vertical\",\n",
    "        )\n",
    "\n",
    "fig = circos.plotfig()\n",
    "\n",
    "# Add legend using dummy traces\n",
    "for query_name, color in comp_name2color.items():\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=[None],\n",
    "            y=None,\n",
    "            mode=\"markers\",\n",
    "            marker=dict(color=parse_color(color), size=14, symbol=\"square\"),\n",
    "            name=query_name,\n",
    "            showlegend=True,\n",
    "        )\n",
    "    )\n",
    "\n",
    "fig.update_layout(\n",
    "    legend=dict(\n",
    "        x=0.5,\n",
    "        y=0.35,\n",
    "        xanchor=\"center\",\n",
    "        yanchor=\"middle\",\n",
    "        font=dict(size=12),\n",
    "        borderwidth=0,\n",
    "    ),\n",
    ")\n",
    "\n",
    "HTML(pio.to_html(fig, include_plotlyjs=\"cdn\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
